File: examples/rag_pipeline.py
"""
Example: RAG Pipeline with GenAI Telemetry
==========================================

This example shows how to instrument a Retrieval-Augmented Generation (RAG)
pipeline with multiple trace decorators.

Prerequisites:
    pip install genai-telemetry openai

Usage:
    export OPENAI_API_KEY="your-api-key"
    python rag_pipeline.py
"""

from typing import List, Dict
from genai_telemetry import (
    setup_telemetry,
    trace_llm,
    trace_embedding,
    trace_retrieval,
    trace_chain,
)

# Initialize telemetry
setup_telemetry(
    workflow_name="rag-pipeline",
    exporter="console"
)

# Simulated document store
DOCUMENTS = [
    {"id": 1, "text": "Paris is the capital of France.", "embedding": [0.1, 0.2, 0.3]},
    {"id": 2, "text": "London is the capital of England.", "embedding": [0.2, 0.3, 0.4]},
    {"id": 3, "text": "Berlin is the capital of Germany.", "embedding": [0.3, 0.4, 0.5]},
    {"id": 4, "text": "Tokyo is the capital of Japan.", "embedding": [0.4, 0.5, 0.6]},
]


@trace_embedding(model="text-embedding-3-small")
def get_query_embedding(query: str) -> List[float]:
    """Generate embedding for the query."""
    # In production, use actual embedding API
    # from openai import OpenAI
    # client = OpenAI()
    # response = client.embeddings.create(
    #     model="text-embedding-3-small",
    #     input=query
    # )
    # return response.data[0].embedding
    
    # Simulated embedding
    return [0.15, 0.25, 0.35]


@trace_retrieval(vector_store="in-memory", embedding_model="text-embedding-3-small")
def retrieve_documents(query_embedding: List[float], top_k: int = 2) -> List[Dict]:
    """Retrieve relevant documents based on embedding similarity."""
    # In production, use actual vector store
    # Simulated retrieval - just return first top_k documents
    return DOCUMENTS[:top_k]


@trace_llm(model_name="gpt-4o-mini", model_provider="openai", extract_content=True)
def generate_answer(query: str, context: str) -> str:
    """Generate answer using LLM with retrieved context."""
    # In production, use actual LLM API
    # from openai import OpenAI
    # client = OpenAI()
    # response = client.chat.completions.create(
    #     model="gpt-4o-mini",
    #     messages=[
    #         {"role": "system", "content": f"Answer based on this context: {context}"},
    #         {"role": "user", "content": query}
    #     ]
    # )
    # return response
    
    # Simulated response
    class MockResponse:
        class Choices:
            class Message:
                content = f"Based on the context, the answer to '{query}' is found in the documents."
            message = Message()
        choices = [Choices()]
        class Usage:
            prompt_tokens = 50
            completion_tokens = 25
        usage = Usage()
    
    return MockResponse()


@trace_chain(name="rag-pipeline")
def rag_query(question: str) -> str:
    """
    Full RAG pipeline:
    1. Generate query embedding
    2. Retrieve relevant documents
    3. Generate answer using LLM
    """
    # Step 1: Generate embedding
    query_embedding = get_query_embedding(question)
    
    # Step 2: Retrieve documents
    relevant_docs = retrieve_documents(query_embedding, top_k=2)
    
    # Step 3: Build context
    context = "\n".join([doc["text"] for doc in relevant_docs])
    
    # Step 4: Generate answer
    answer = generate_answer(question, context)
    
    return answer


if __name__ == "__main__":
    print("=" * 60)
    print("RAG Pipeline Example")
    print("=" * 60)
    
    question = "What is the capital of France?"
    print(f"\nQuestion: {question}")
    
    answer = rag_query(question)
    print(f"\nAnswer: {answer}")
    
    print("\n" + "=" * 60)
    print("Check the console output above for the full trace!")
    print("=" * 60)

File: js/examples/rag_pipeline.ts
/**
 * RAG Pipeline Example
 * 
 * This example demonstrates how to trace a complete RAG pipeline
 * including embeddings, retrieval, and LLM calls.
 */

import { 
  setupTelemetry, 
  traceLLM, 
  traceEmbedding, 
  traceRetrieval,
  traceChain,
  getTelemetry
} from 'genai-telemetry';

// Initialize telemetry
setupTelemetry({
  workflowName: 'rag-pipeline',
  exporter: 'console',
  // Or use multiple exporters:
  // exporter: [
  //   { type: 'splunk', url: 'http://splunk:8088', token: 'xxx' },
  //   { type: 'console' }
  // ]
});

// Simulated vector store
const documents = [
  { id: 1, content: 'Paris is the capital of France.', embedding: [] },
  { id: 2, content: 'London is the capital of the United Kingdom.', embedding: [] },
  { id: 3, content: 'Tokyo is the capital of Japan.', embedding: [] },
  { id: 4, content: 'The Eiffel Tower is located in Paris.', embedding: [] },
  { id: 5, content: 'Big Ben is a famous clock tower in London.', embedding: [] },
];

// Simulated embedding function
const generateEmbedding = traceEmbedding(
  async (text: string): Promise<{ data: [{ embedding: number[] }], usage: { total_tokens: number } }> => {
    // Simulate API delay
    await new Promise(resolve => setTimeout(resolve, 50));
    
    // Return mock embedding
    return {
      data: [{ embedding: Array(1536).fill(0).map(() => Math.random()) }],
      usage: { total_tokens: text.split(' ').length }
    };
  },
  { model: 'text-embedding-3-small' }
);

// Simulated vector search
const searchDocuments = traceRetrieval(
  async (query: string): Promise<typeof documents> => {
    // Simulate search delay
    await new Promise(resolve => setTimeout(resolve, 100));
    
    // Simple keyword search (in real app, use vector similarity)
    const keywords = query.toLowerCase().split(' ');
    const results = documents.filter(doc => 
      keywords.some(kw => doc.content.toLowerCase().includes(kw))
    );
    
    return results.slice(0, 3);
  },
  { vectorStore: 'in-memory', embeddingModel: 'text-embedding-3-small' }
);

// Simulated LLM call
const generateAnswer = traceLLM(
  async (prompt: string): Promise<{
    choices: [{ message: { content: string } }],
    usage: { prompt_tokens: number, completion_tokens: number }
  }> => {
    // Simulate API delay
    await new Promise(resolve => setTimeout(resolve, 200));
    
    // Mock response
    return {
      choices: [{
        message: {
          content: `Based on the context provided, here is my answer: ${prompt.slice(0, 100)}...`
        }
      }],
      usage: {
        prompt_tokens: prompt.split(' ').length,
        completion_tokens: 50
      }
    };
  },
  { modelName: 'gpt-4o', modelProvider: 'openai' }
);

// Complete RAG pipeline
const ragPipeline = traceChain(
  async (question: string): Promise<string> => {
    const telemetry = getTelemetry();
    
    console.log(`\nProcessing question: ${question}`);
    
    // Step 1: Generate embedding for the query
    console.log('1. Generating query embedding...');
    await generateEmbedding(question);
    
    // Step 2: Search for relevant documents
    console.log('2. Searching for relevant documents...');
    const relevantDocs = await searchDocuments(question);
    console.log(`   Found ${relevantDocs.length} relevant documents`);
    
    // Step 3: Build context from retrieved documents
    const context = relevantDocs.map(doc => doc.content).join('\n');
    
    // Step 4: Generate answer using LLM
    console.log('3. Generating answer...');
    const prompt = `Context:\n${context}\n\nQuestion: ${question}\n\nAnswer:`;
    const response = await generateAnswer(prompt);
    
    return response.choices[0].message.content;
  },
  { name: 'rag-qa-pipeline' }
);

// Main function
async function main() {
  console.log('='.repeat(60));
  console.log('RAG Pipeline Example');
  console.log('='.repeat(60));

  const questions = [
    'What is the capital of France?',
    'Tell me about famous landmarks in London.',
    'What do you know about Tokyo?',
  ];

  for (const question of questions) {
    console.log('\n' + '-'.repeat(60));
    const answer = await ragPipeline(question);
    console.log('\nAnswer:', answer);
  }

  console.log('\n' + '='.repeat(60));
  console.log('Done! Check your telemetry backend for the complete trace.');
}

main().catch(console.error);
